[WIP] Add Operator Observability Best Practices (#5975)
operator-framework/operator-sdk: website/content/en/docs/best-practices/observability-best-practices.md

---

## Operator Observability Best Practices

In this document we will outline what operators require in order to meet the "Deep Insights" capability level and provide best practices and examples for creating metrics, [recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules) and alerts. It is based on the general guidelines in [Operator Capability Levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/).

**Note:** For technical documentation of how to add metrics to your operator, please read the [Metrics](https://book.kubebuilder.io/reference/metrics.html) section of the Kubebuilder documentation.

### Deep Insights capability level requirements

1. **Health and Performance metrics** - for all of the operator components  
    1.1. Metrics should be implemented based on the guidelines below.  
    1.2. **Metrics Documentation** - All metrics should have documentation.  
    1.3. **Metrics Tests** - Metrics should include test that verify that they exist and that their value is correct.  
2. **Alerts** for when things are not working as expected for each of the operator's components  

>     2.1  Alerts should br implemented based on the guidelines below.  

Reviewer: Alerts should be (Typo)

    2.3. **Alerts Tests** - Alerts should include E2E Testing and unit tests.  
3. **Events** - Custom Resources should emit custom events for the operations taking place.

### Metrics Guidelines

#### Metrics Naming
Your operator metrics should align with the Kubernetes metrics names.

Your operator users should get the same experience when searching for a metric across Kubernetes operators, resources and custom resources.
1. Check if a similar Kubernetes metric, for node, container or pod, exists and try to align to it.
2. The metrics search list in the Prometheus and Grafana UI is sorted in alphabetical order.
When searching for a metric, it should be easy to identify metrics that are related to a specific operator.
That is why we recommend that your operator metrics name to follow this format:
`Operator name` prefix + the `Sub operator name` or `entity` + `metric name` based on the [Prometheus naming conventions](https://prometheus.io/docs/practices/naming/).

**Note:** In Kubernetes metrics are separated like this:
- node_network_**receive**_packets_total

> - node_network_**transmit**_packets_total
> In this example based on `receive` and `transmit`.

Reviewer 2: ```suggestion
- node_network_**transmit**_packets_total

In this example based on `receive` and `transmit`.
```
Without newline in between, both lines are rendered as one.

- kubevirt_vmi_network_**transmit**_bytes_total
- kubevirt_migrate_vmi_**data_processed**_bytes
- kubevirt_migrate_vmi_**data_remaining**_bytes

#### Metrics `Help` message

Your operator metrics `help` message should include the following details:
- What does this metric measure?
- What is the unit of measurement? See the [Prometheus base units](https://prometheus.io/docs/practices/naming/#base-units) and [Understanding metrics types](https://prometheus.io/docs/tutorials/understanding_metric_types/#types-of-metrics)
- What does the output mean?
When creating a new metric or recording rule that reports a resource like a `pod` or a `container` name, please make sure that the `namespace` is included, in order for it to be uniquely identified.
**Note:** Usually the `namespace` label is populated via service discovery, but there can be cases where it should be added explicitly, for example if you report a metric for a resource that doesn't live in the same namespace as the operator or when creating a recording rules.
- What is the [metric type](https://prometheus.io/docs/concepts/metric_types/#metric-types)? For example: `Gauge`/`Counter`/`Histogram` etc.

> - What unique labels does the metric use? if applicable. Please note (https://prometheus.io/docs/practices/naming/#labels).

Reviewer 2: This should be a separate section `Metrics Labels`.

      15 minutes. There are many potential causes of OOM errors, however issues
      on a specific node or containers breaching their limits is common.
      summary: Containers are being killed due to OOM
  expr: sum(max by(namespace, container, pod) (increase(kube_pod_container_status_restarts_total[12m]))
    and max by(namespace, container, pod) (kube_pod_container_status_last_terminated_reason{reason="OOMKilled"}) == 1) > 5
  for: 15m
  labels:
    namespace: kube-system
    severity: info
```

This alert fires if multiple containers have been terminated due to out of
memory conditions in the last 15 minutes.  This is something the administrator
should be aware of, but may not require immediate action.

### Alerts, Metrics and Recording Rules Tests

1. Add tests for alerts that validate that:
   - Each alert includes all mandatory fields.
   - Each `runbook_url` link is valid.
   - Each alert that includes a `pod` or a `container` also includes the `namespace`.
2. Add e2e tests that inspect the alerts during upgrade and make sure that the alerts don’t fire when they shouldn’t (Zero noise).
3. Add tests for metrics/recording rules that validate that:

>    - Metric / Recording rule exists
>    - Metric / Recording rule value is as expected
>    - Metric / Recording rule name follows the best practices guideline

Reviewer 2: ```suggestion
   - Metric / Recording rule exists.
   - Metric / Recording rule value is as expected.
   - Metric / Recording rule name follows the best practices guideline.
```

---
title: "Operator Observability Best Practices"
linkTitle: "Observability Best practices"
weight: 6
description: This guide describes the best practices concepts for adding Observability to operators.
---

## Operator Observability Best Practices

> In this document we will outline what operators require in order to meet the "Deep Insights" capability level and provide best practices and examples for creating metrics, recording rules and alerts. It is based on the general guidelines in [Operator Capability Levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/).

Reviewer 3: What do you mean with `recording rules`?

Reviewer 4: I guess this may refer to [prometheus recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules).
In general, the prometheus user would compose various `prom_ql` to calculate the status of the object under monitoring.
Well, some `prom_ql` maybe frequently applied and computed, others may be "computationally expensive".
A `recording rule` is like an alias to one of a `prom_ql` that is frequently used, so that prometheus will "record" the value of the query before you actually call it, so that save time.

> 

Reviewer 3: This doc only speaks about metrics and alerts. Am I right?
So, I think the title here and overview needs to be specific to metrics and alerts. 

OR 

We need to cover all aspects of how to achieve the level which is not only about metrics and alerts, see: https://sdk.operatorframework.io/docs/overview/operator-capabilities/

> **Note:** For technical documentation of how to add metrics to your operator, please read the [Metrics](https://book.kubebuilder.io/reference/metrics.html) section of the Kubebuilder documentation.

Reviewer 3: @sradco 

We have the level capabilities described here: https://sdk.operatorframework.io/docs/overview/operator-capabilities/.

For this one, the criteria are:

- Metrics
- Alerts,
- Logs
- Processing and workload analysis

However, by reading this doc I do think we are covering all and its content seems more related to the metrics/alerts naming conventions and good practices, So, what do you think about re-wording a little here?

### Deep Insights capability level requirements

1. **Health and Performance metrics** for all of the operator components - Implemented based on the guidelines below.
2. **Metrics Documentation** - All metrics should have documentation.
3. **Alerts** for when things are not working as expected for each of the operator's components - Implemented based on the guidelines below.

> 4. **Alerts Runbooks** - Each alert MUST include a runbook_url annotation and a runbook.

Reviewer 4: I guess this refers to the link for runbook. For a prometheus alert, it can be attached with such `runbook_url` so that when the alert triggered, the people on-call can go to the link and see the procedure to resolve the issue.

e.g.:
<img width="1407" src="https://user-images.githubusercontent.com/18136486/205117624-f2fdcfa0-f977-480a-94d1-7060980171e0.png">

> 4. **Alerts Runbooks** - Each alert MUST include a runbook_url annotation and a runbook.

Author: Yes, I updated it to be optional.

5. **Alerts and Metrics Tests** - E2E Testing for metrics and alerts and unit tests for alerts.

> 6. **Events** - Custom Resources MUST emit custom events for the operations taking place.

Reviewer 3: We should NOT emit events for all Operations. If you raise too many events it brings bad UX experiences for those that are consuming the solutions on the cluster and they will not have attention to what they actually must be notified. 

Events should be raised in certain circumstances only and following this guideline acctually is not a good practice. See that the Kubernetes APIs convention describes when use them here: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#events

> Events are complementary to status information, since they can provide some historical information about status and occurrences in addition to current or previous status. **Generate events for situations users or administrators should be alerted about.**

See that is recommended properly follow the Kubernetes API convention and use conditional. status. Therefore, the cluster admin will be able to check the CR status and know the status of the desired state on the cluster. See: 
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

@sradco ^

> 7. **Metering** - Operator leverages Operator Metering.

Reviewer 3: ```suggestion
```

We can recommend/suggest using this solution to achieve some goal but it is not a pre-requirement to achieve the level. Also, why use this solution would be a recommendation? What advantages it can bring? What complexities to achieve the level it solves?

Author: @camilamacedo86 +1. Ill add this info

   **Note:** Metering isn't mandatory at this point in order to meet the "Deep Insights" capability level requirments.

> 

Reviewer 3: @sradco the requirements are defined in: https://sdk.operatorframework.io/docs/overview/operator-capabilities/
and does not seems to be the same. For example Metrics Tests are not a requirement.

Author: @camilamacedo86 Both tests and runbooks are part of the metrics and alerts. I will refactor the requirements and add them under "alerts" and "metrics".

Author: Done

### Metrics Guidelines

> 1. Metrics `Help` message should be verbose, since it can be used to create auto generated documentation, Like its done here for example [KubeVirt metrics](https://github.com/kubevirt/kubevirt/blob/main/docs/metrics.md) and generated by [KubeVirt metrics doc generator](https://github.com/kubevirt/kubevirt/blob/main/tools/doc-generator/doc-generator.go).

Reviewer 3: I think you would like to say here that the help should be provided and that it uses to create a documentation

> 1. Metrics `Help` message should be verbose, since it can be used to create auto generated documentation, Like its done here for example [KubeVirt metrics](https://github.com/kubevirt/kubevirt/blob/main/docs/metrics.md) and generated by [KubeVirt metrics doc generator](https://github.com/kubevirt/kubevirt/blob/main/tools/doc-generator/doc-generator.go).

Reviewer 3: ```suggestion
1. Metrics `Help` message should be meaningful since it can be used to create auto-generated documentation Like it's done here for example [KubeVirt metrics](https://github.com/kubevirt/kubevirt/blob/main/docs/metrics.md) and generated by [KubeVirt metrics doc generator](https://github.com/kubevirt/kubevirt/blob/main/tools/doc-generator/doc-generator.go).
```

I think you would like to say here that the help should be provided and that it can be used to auto-generate documentation. Therefore, I think the word "verbose" does not fit well.. maybe meaningful? See that we can have a very verbose text that is not meaningful at all. Therefore, an explanation/description does not need to be verbose to be good.

Reviewer 3: About the examples, have we some examples of it in the Prometheus documentation itself? Or is its official good practices and guidelines docs?

Author: @camilamacedo86 examples for the help message you mean?

Author: @camilamacedo86  The Prometheus guidelines are not that verbose. We added more details.
We can add later examples from the Memchached operator. I'll make sure that Assaf adds a full description to the new metrics.

> 

Reviewer 3: @sradco one thing that would be great to add here or in a third doc for we link is some examples about how to work with the events ( how to raise an event ) and how to deal with status conditions. 

## How to raise an event:

a) You need to pass the recorder when you set up the recorder for the controller when the event will be raised see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/main.go#L103

b) You need to add the makers to add the RBAC permissions to allow your Operator/controller raise the event see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L66 and run make manifests

c) You can check en example of the event been called in : https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L299-L303

## How to work with status:

a) It is recommended use status conditionals see an example: https://github.com/kubernetes-sigs/kubebuilder/blob/7cd3532662567e0a7568415e271f0b29cece202c/testdata/project-v4-with-deploy-image/api/v1alpha1/busybox_types.go#L57-L64

b) Then, you can update the status as it is done in the reconciliation as example here: https://github.com/kubernetes-sigs/kubebuilder/blob/7cd3532662567e0a7568415e271f0b29cece202c/testdata/project-v4-with-deploy-image/controllers/busybox_controller.go#L204-L207

c) that you also need to add the marker to give the permissions to manage the status, see: https://github.com/kubernetes-sigs/kubebuilder/blob/7cd3532662567e0a7568415e271f0b29cece202c/testdata/project-v4-with-deploy-image/controllers/busybox_controller.go#L64

The above code examples are part of the deploy-image plugin, more info: https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html

Author: @camilamacedo86  I believe we should add this data to https://book.kubebuilder.io/reference/. We should create an "observability"  section and have alerts, metrics, events and logs under it.

Reviewer 3: I think these both would be great there. 
We can create one section for each one. Like :: events and other status conditions then we can link.

The `help` message should include the following details:
- What does this metric measure
- What is the unit of measurement. See the [Prometheus base units](https://prometheus.io/docs/practices/naming/#base-units) and [Understanding metrics types](https://prometheus.io/docs/tutorials/understanding_metric_types/#types-of-metrics)
- What does the output mean.

> When creating a new metric or recording rule that reports a resource like a ‘pod’ or a ‘container’ name, please make sure that the `namespace` is included, in order for it to be uniquely identified. 

Reviewer 3: A consumer can install the Operator in any namespace so how the author will define the namespace beforehand?

Author: @camilamacedo86 The namespace label is populated via service discovery in general. The exception would be if you report a metric for a resource that doesn't live in the same namespace as the operator. Also in recording rules you need to make sure to also include the `namespace` label in the calculation. See the "Note" that I added below this line.

Reviewer 3: @sradco 
What do you mean by `The namespace label is populated via service discovery in general`?
Could you please provide an example?

Author: @camilamacedo86  The service discovery knows the namespace associated to the pod/container and usually they are reported as part of the metrics without the need to specify it in the metric. But there are a few exceptions that this is not automatic, like when creating a recording rule. A recording rule is reported like a metric but doesn't include a "Type" (like gauge/counter/etc.) and doesn't include all fields from the respected metrics that it is based on. In recording rules the developer should include the namespace specifically when there is a resource that can have the same name in different namespaces.

Let me know if this is clear and feel free to suggest a better wording for this.

Reviewer 4: @sradco Do you mean [labelling `namespace` on metrics](https://prometheus.io/docs/practices/naming/#labels) instead of k8s resources?

> **Note:** Usually the ‘namespace’ label is populated via service discovery, but there can be cases where it should be added explicitly, usually this can happen for recording rules.

Reviewer 3: Also, see that Kubernetes has a doc about the label's good practices and the namespace is not part of them, see: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/

Why we would need to have a label with the namespace?

Author: I hit this specific issue. Its related specifically for k8s. If you want to identify a container/pod you must have its namespace since the same name of a pod/container can live in more than one namespace. To get a 1:1 match you must have the namespace.

Reviewer 4: @sradco are you suggesting labelling "namespaces" [on metrics](https://prometheus.io/docs/practices/naming/#labels)? (where the user does not necessarily need to add label/annotation of namespace on K8s resources.)

Author: @Kavinjsir If a metric includes a pod or other resource that is tied to a namspace and its name be the same name in different namespaces, then the developer should make sure the metric includes the namespace so that the user can identify the correct resource in question.

In order to easily identify your operator recording rules, their name should follow the same naming guidelines as the metrics.

### Alerts Guidelines
Clear and actionable alerts are a key component of a smooth operational
experience. Ensuring we have clear and concise guidelines for developers and
administrators creating new alerts will result in a better
experience for end users.

There should be a clear guidance aligning critical, warning, and info
alert severities with expected outcomes across components, to avoide alerts fatigue for administrators.
There must be an acceptance criteria for critical alerts.

#### Recommended Reading

A list of references on good alerting practices:

* [Google SRE Book - Monitoring Distributed Systems](https://sre.google/sre-book/monitoring-distributed-systems/)
* [Prometheus Alerting Documentation](https://prometheus.io/docs/practices/alerting/)
* [Alerting for Distributed Systems](https://www.usenix.org/sites/default/files/conference/protected-files/srecon16europe_slides_rabenstein.pdf)

#### Alert Ownership

Individual teams are responsible for writing and maintaining alerting rules for

> their components, i.e. their operators and operands.

Reviewer 3: Teh audience here is Operator Authors so I think we need re-wording a little

Author: Why. What I meant here is that each operator introduces and maintains their own alerts. How would you rephrase and is this section needed?

Reviewer 3: I do think that we should use the word "teams" here. It would be "Operator Authors" as we can see in the other docs.

Author: Done

Teams should also take into consideration how their components interact with
existing monitoring and alerting.  As an example, if your operator deploys a
service which creates one or more `PersistentVolume` resources, and these
volumes are expected to be mostly full as part of normal operation, it's likely
that this will cause unnecessary `KubePersistentVolumeFillingUp` alerts to fire.
You should work with the monitoring team to find a solution to avoid triggering
these alerts if they are not actionable.

> 

Reviewer 3: So, I understand that we need to alert any scenario that:
a) Cannot be sorted out programmatically by the operator
b) That is not be expected to check with frequency 
c) That can cause a risk like high usage of resources.

Am I right? if so, could we try to formulate to be a little more directly? WDYT?

Author: As an operator owner that reads this document I want to explain about naming conventions, labels, runbooks and tests that operators need.
I don't think we should explain here about when to alert, that is up to the operator.
There are a few guideline documents listed above in the "Recommended Reading" section for that.

Author: @camilamacedo86 WDYT?

description annotations, and it links to a runbook with information on
investigating and resolving the issue.

The group of critical alerts should be small, very well defined, highly
documented, polished and with a high bar set for entry.  This includes a
mandatory review of a proposed critical alert by the Red Hat SRE team.

##### Warning Alerts

TL/DR: The vast majority of alerts should use the severity.  Issues at the
warning level should be addressed in a timely manner, but don't pose an
immediate threat to the operation of the cluster as a whole.

Timeline:  ~60 minutes

If your alert does not meet the criteria in "Critical Alerts" above, it belongs
to the warning level or lower.

Use warning level alerts for reporting conditions that may lead to inability to
deliver individual features of the cluster, but not service for the cluster as a
whole. Most alerts are likely to be warnings. Configure warning level alerts so
that they do not fire until components have sufficient time to try to recover
from the interruption automatically. Expect users to be notified of a warning,
but for them not to respond with corrective action immediately.

> Example warning alert: [ClusterNotUpgradeable](https://github.com/openshift/cluster-version-operator/blob/513a2fc/install/0000_90_cluster-version-operator_02_servicemonitor.yaml#L68-L76)

Reviewer 3: Have we any examples that do not come from OCP?

Author: I would like to replace these examples with memchached alerts, once we have them in place. That is the next item for Assaf to work on.

Author: @camilamacedo86 ^ woul it make sense to use memchached examples here?

  for: 15m
  labels:
    namespace: kube-system
    severity: info
```

This alert fires if multiple containers have been terminated due to out of
memory conditions in the last 15 minutes.  This is something the administrator
should be aware of, but may not require immediate action.

### Alerts, Metrics and Recording Rules Tests

1. Add tests for alerts that validate that:
   - Each alert includes all mandatory fields.
   - Each `runbook_url` link is valid.
   - Each alert that includes a `pod` or a `container` also includes the `namespace`.
2. Add e2e tests that inspect the alerts during upgrade and make sure that the alerts don’t fire when they shouldn’t (Zero noise).
3. Add tests for metrics/recording rules that validate that:
   - Metric / Recording rule exists
   - Metric / Recording rule value is as expected
   - Metric / Recording rule name follows the best practices guideline

### Test Plan

Automated tests enforcing acceptance criteria for critical alerts and basic style linting will be added to the [openshift/origin](https://github.com/openshift/origin) end-to-end test suite.
The monitoring team will work with anyone shipping existing critical alerts that

> don't meet these criteria in order to resolve the issue before enabling the tests.

Reviewer 3: This one does not fit well here. We should not add anything that is specific to OCP.

---
title: "Operator Observability Best Practices"

> linkTitle: "Observability Best practices"

Author: Fixed

All operators start small. This separation will help you, as a developer, with easier maintenance of both your operator core code and the monitoring code and for other stakeholders to understand your monitoring code better.

### Metrics Guidelines

#### Metrics Naming
Kubernetes components emit metrics in [Prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/). This format is structured plain text, designed so that people and machines can both read it.

Your operator users should get the same experience when searching for a metric across Kubernetes operators, resources and custom resources.
1. Check if a similar Kubernetes metric, for node, container or pod, exists and try to align to it.
2. The metrics search list, in the Prometheus, Grafana UI and even in the /metrics end point, is sorted in alphabetical order.
When searching for a metric, it should be easy to identify metrics that are related to a specific operator.
That is why we recommend that your operator metrics name will follow this format:

> `Operator name` prefix + the `Sub operator name` or `entity` + `metric name` based on the [Prometheus naming conventions](https://prometheus.io/docs/practices/naming/). For example, in the [memcached-operator](https://github.com/operator-framework/operator-sdk/blob/0d2fa86f0d3cc92c4672cb9e1d246efaefcf7ced/testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/metrics.go#L14).

Reviewer: ```suggestion
`operator name` prefix + `sub-operator name` or `entity` + `metric name` based on the [Prometheus naming conventions](https://prometheus.io/docs/practices/naming/). For example, in the [memcached-operator](https://github.com/operator-framework/operator-sdk/blob/0d2fa86f0d3cc92c4672cb9e1d246efaefcf7ced/testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/metrics.go#L14).
```

**Note:** In [Prometheus Node Exporter](https://github.com/prometheus/node_exporter) metrics are separated like this:
- node_network_**receive**_packets_total
- node_network_**transmit**_packets_total  
 
In this example, based on `receive` and `transmit`.

Please follow the same principle and don't put similar metrics details as labels, so the user experience would be fluent.  
Example for this in an operator:
- kubevirt_vmi_network_**receive**_errors_total
- kubevirt_vmi_network_**transmit**_bytes_total
- kubevirt_migrate_vmi_**data_processed**_bytes
- kubevirt_migrate_vmi_**data_remaining**_bytes

3. Your metric suffix should indicate the metric unit. For better compatibility, [Prometheus base units](https://prometheus.io/docs/practices/naming/#base-units) should be used.

> 4. Prometheus supports four [metric types](https://prometheus.io/docs/concepts/metric_types/#metric-types). `Gauge`,`Counter`,`Histogram` and `Summery`. You can read more about the different types here, [Understanding metrics types](https://prometheus.io/docs/tutorials/understanding_metric_types/#types-of-metrics).  

Reviewer: ```suggestion
4. Prometheus supports four [metric types](https://prometheus.io/docs/concepts/metric_types/#metric-types). `Gauge`,`Counter`,`Histogram` and `Summary`. You can read more about the different types here, [Understanding metrics types](https://prometheus.io/docs/tutorials/understanding_metric_types/#types-of-metrics).  
```

Author: Fixed

The most common types are:
 - `Counter` - Value can only increase or reset.
 - `Gauge` Value can be increased and decreased as needed.
5. `_total` suffix should be used for accumulating count. If your metrics has labels with high cardinality, like `pod`/`container` it usually means that you can aggregate it more, thus it will not require `_total` suffix.

### Prometheus Labels
[Prometheus labels](https://prometheus.io/docs/practices/naming/#labels) are used to differentiate the characteristics of the thing that is being measured.

1. **Important** - Be cautious when adding labels to metrics. Labels can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values. Note: There are still cases when we will still need to have a high cardinality label like the `pod name`, but try to keep this to the minimum.

> 2. When creating a new metric, recording rule or alert, that reports a resource like a `pod` or a `container` name, make sure that the `namespace` is included, in order to be able to uniquely identify it.

Reviewer: ```suggestion
2. When creating a new metric, recording rule or alert, that reports a resource like a `pod` or a `container`, make sure that the `namespace` is included, in order to be able to uniquely identify it.
```

Author: Fixed

#### Metrics `Help` message

Your operator metrics `help` message should include the following details:
- What does this metric measure?
- What does the output mean?
- What important labels does the metric use? (Optional. If applicable).

> The `Help` message can be used to create auto-generated documentation, Like it's done here for example in [KubeVirt](https://github.com/kubevirt/kubevirt/blob/main/docs/metrics.md) and generated by the [KubeVirt metrics doc generator](https://github.com/kubevirt/kubevirt/blob/main/tools/doc-generator/doc-generator.go).

Reviewer: ```suggestion
The `Help` message can be used to create auto-generated documentation, like it's done in [KubeVirt](https://github.com/kubevirt/kubevirt/blob/main/docs/metrics.md) and generated by the [KubeVirt metrics doc generator](https://github.com/kubevirt/kubevirt/blob/main/tools/doc-generator/doc-generator.go).
```

Author: Fixed

> We recommend to auto-generated metrics documentation and save it in your operator repository, to a location like to `/docs/monitoring/`, so that the users can find the information about your operator metrics easily. 

Reviewer: ```suggestion
We recommend to auto-generated metrics documentation and save it in your operator repository, to a location like `/docs/monitoring/`, so that the users can find the information about your operator metrics easily. 
```

Reviewer: In Memcached operator the metrics.md location defined in Makefile.

Author: Fixed

#### Prometheus Recording Rules Naming
As per [Prometheus](https://prometheus.io/docs/prometheus) documentation, [Recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules) allow you to pre-compute frequently needed or computationally expensive expressions and save their result as a new set of time series.

**Note:** The Prometheus recording rules appear in Prometheus UI as metrics.

> In order to easily identify your operator recording rules, their name should usually follow the same naming guidelines as the metrics.

Reviewer: ```suggestion
In order to easily identify your operator recording rules, their names should usually follow the same naming guidelines as the metrics.
```

Author: Fixed

### Prometheus Alerts Guidelines
Clear and actionable alerts are a key component of a smooth operational
experience and will result in a better experience for the end users.

> The following guidances aim to align alert naming, severities, labels, etc.

Reviewer: ```suggestion
The following guidances aim to align alert naming, severities, labels, etc.,
```

Author: Fixed

Use these sparingly, and consider instead reporting this information via Kubernetes events.

Example info alert: [MultipleContainersOOMKilled](https://github.com/openshift/cluster-monitoring-operator/blob/79cdf68/assets/cluster-monitoring-operator/prometheus-rule.yaml#L326-L338)

```yaml
- alert: MultipleContainersOOMKilled
  annotations:
    description: Multiple containers were out of memory killed within the past
      15 minutes. There are many potential causes of OOM errors, however issues
      on a specific node or containers breaching their limits is common.
      summary: Containers are being killed due to OOM
  expr: sum(max by(namespace, container, pod) (increase(kube_pod_container_status_restarts_total[12m]))
    and max by(namespace, container, pod) (kube_pod_container_status_last_terminated_reason{reason="OOMKilled"}) == 1) > 5
  for: 15m
  labels:
    namespace: kube-system
    severity: info
```

This alert fires if multiple containers have been terminated due to out of memory conditions in the last 15 minutes.  
This is something the administrator should be aware of, but may not require immediate action.

> ### Alerts, Metrics and Recording Rules Tests

Reviewer: consider mentioning this section at the end of the Metrics Guidelines, as a remainder to include tests.

Author: Added comments to metrics and recording rules sections

### Operator Observability Recommended Components

1. **Health and Performance metrics** - for all of the operator components  
    1.1. Metrics should be implemented based on the guidelines below.  
    1.2. **Metrics Documentation** - All metrics should have documentation.  
    1.3. **Metrics Tests** - Metrics should include test that verify that they exist and that their value is correct.  
2. **Alerts** for when things are not working as expected for each of the operator's components  
    2.1  Alerts should be implemented based on the guidelines below.  
    2.2. **Alerts Runbooks** - Each alert can include a `runbook_url` annotation and an alert runbook that describes it. See additional details below.  
    2.3. **Alerts Tests** - Alerts should include E2E Testing and unit tests.  
3. **Events** - Custom Resources should emit custom events for the operations taking place.

Additional component would be `Dashboards` and `Logs`, which are not addressed in this document at this point.  

### Operators Observability General Guidelines
**Important:** It is highly recommended to separate your monitoring code from your core operator code.

> We recommend to create a dedicated `/monitoring` subfolder that will include all the code of the "Operator Observability Recommended Components" that are outlined above. For example, in the [memcached-operator](https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v4-alpha/monitoring/memcached-operator/monitoring).

Author: @varshaprasad96  I added for each section where we recommend to add the code.
Once we have all the examples the structure would be very visible from the link.

---
title: "Operator Observability Best Practices"
linkTitle: "Observability Best practices"
weight: 6
description: This guide describes the best practices concepts for adding Observability to operators.
---

## Operator Observability Best Practices

> In this document we provide best practices and examples for creating metrics, [recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules) and alerts. It is based on the general guidelines in [Operator Capability Levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/).

Reviewer 5: ```suggestion
In this document, we provide best practices and examples for creating metrics, [recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules) and alerts. It is based on the general guidelines in [Operator Capability Levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/).
```

**Note:** For technical documentation of how to add metrics to your operator, please read the [Metrics](https://book.kubebuilder.io/reference/metrics.html) section of the Kubebuilder documentation.

### Operator Observability Recommended Components

1. **Health and Performance metrics** - for all of the operator components  
    1.1. Metrics should be implemented based on the guidelines below.  
    1.2. **Metrics Documentation** - All metrics should have documentation.  
    1.3. **Metrics Tests** - Metrics should include test that verify that they exist and that their value is correct.  
2. **Alerts** for when things are not working as expected for each of the operator's components  
    2.1  Alerts should be implemented based on the guidelines below.  
    2.2. **Alerts Runbooks** - Each alert can include a `runbook_url` annotation and an alert runbook that describes it. See additional details below.  
    2.3. **Alerts Tests** - Alerts should include E2E Testing and unit tests.  
3. **Events** - Custom Resources should emit custom events for the operations taking place.

> Additional component would be `Dashboards` and `Logs`, which are not addressed in this document at this point.  

Reviewer 4: What about adding `tracing` here for future documentation?
(It may not be necessary for all operators, well in terms of "observability", this can be covered.
One use case may be: observing webhooks)

Author: Added. Thanks.

> Additional component would be `Dashboards` and `Logs`, which are not addressed in this document at this point.  

Reviewer 5: ```suggestion
Additional components would be `Dashboards` and `Logs`, which are not addressed in this document at this point.  
```

### Operators Observability General Guidelines
**Important:** It is highly recommended to separate your monitoring code from your core operator code.  
We recommend to create a dedicated `/monitoring` subfolder that will include all the code of the [Operator Observability Recommended Components](#operator-observability-recommended-components), that are outlined above. For example, in the [memcached-operator](https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v4-alpha/monitoring/memcached-operator/monitoring).

> In your core operator code only call the functions the functions that will update the metrics value from you desired location. For example, in the [memcached-operator](https://github.com/operator-framework/operator-sdk/blob/367bd3597c30607099aa73637f5286f7120b847a/testdata/go/v3/monitoring/memcached-operator/controllers/memcached_controller.go#L242).

Reviewer 5: ```suggestion
In your core operator code only call the functions the functions that will update the metrics value from your desired location. For example, in the [memcached-operator](https://github.com/operator-framework/operator-sdk/blob/367bd3597c30607099aa73637f5286f7120b847a/testdata/go/v3/monitoring/memcached-operator/controllers/memcached_controller.go#L242).
```

> All operators start small.This separation will help you, as a developer, with easier maintenance of both your operator core code and the monitoring code and for other stakeholders to understand your monitoring code better.

Reviewer 5: nit: Missing white space

```suggestion
All operators start small. This separation will help you, as a developer, with easier maintenance of both your operator core code and the monitoring code and for other stakeholders to understand your monitoring code better.
```

Your operator metrics should align with the Kubernetes metrics names.

Your operator users should get the same experience when searching for a metric across Kubernetes operators, resources and custom resources.
1. Check if a similar Kubernetes metric, for node, container or pod, exists and try to align to it.
2. The metrics search list in the Prometheus and Grafana UI is sorted in alphabetical order.
When searching for a metric, it should be easy to identify metrics that are related to a specific operator.
That is why we recommend that your operator metrics name to follow this format:
`Operator name` prefix + the `Sub operator name` or `entity` + `metric name` based on the [Prometheus naming conventions](https://prometheus.io/docs/practices/naming/). For example, in the [memcached-operator](https://github.com/operator-framework/operator-sdk/blob/0d2fa86f0d3cc92c4672cb9e1d246efaefcf7ced/testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/metrics.go#L14).

> **Note:** In Kubernetes metrics are separated like this:

Reviewer 4: Can we have a link of the reference to those `kuberntes metrics`?

Author: Its actually the Prometheus node exporter metrics. I updated the text and added the link.

- node_network_**receive**_packets_total
- node_network_**transmit**_packets_total  
 
In this example based on `receive` and `transmit`.

Please follow the same principle and don't put similar metrics details as labels, so the user experience would be fluent.
Example for this in an operator:
- kubevirt_vmi_network_**receive**_errors_total
- kubevirt_vmi_network_**transmit**_bytes_total
- kubevirt_migrate_vmi_**data_processed**_bytes
- kubevirt_migrate_vmi_**data_remaining**_bytes

#### Metrics `Help` message

> 

Reviewer 4: If this is talking about the `help` field defined in a certain metric, then I'm not sure if it is better to be short and efficient?
That is to say: no need to specify: `metric type`, `unique labels`.

See[ OpenMetrics Ref](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#help)

Author: I updated the test. I did leave the `unique labels`, since they can be important for understanding the metric, but its optional.

Reviewer 4: I like the idea to describe `important labels`!

experience and will result in a better experience for the end users.

The following guidances aim to align alert naming, severities, labels, etc.
in order to avoid alerts fatigue for administrators.

#### Recommended Reading

A list of references on good alerting practices:

* [Google SRE Book - Monitoring Distributed Systems](https://sre.google/sre-book/monitoring-distributed-systems/)
* [Prometheus Alerting Documentation](https://prometheus.io/docs/practices/alerting/)
* [Alerting for Distributed Systems](https://www.usenix.org/sites/default/files/conference/protected-files/srecon16europe_slides_rabenstein.pdf)

#### Alert Ownership

Individual operator authors are responsible for writing and maintaining alerting rules for
their components, i.e. their operators and operands.

Operator authors should also take into consideration how their components interact with
existing monitoring and alerting.

As an example, if your operator deploys a service which creates one or more `PersistentVolume` resources,
and these volumes are expected to be mostly full as part of normal operation, it's likely
that this will cause unnecessary `KubePersistentVolumeFillingUp` alerts to fire.

> You should work to find a solution to avoid triggering these alerts if they are not actionable.

Reviewer 4: This seems relative to [inhibition](https://prometheus.io/docs/alerting/latest/alertmanager/#inhibition) and [silences](https://prometheus.io/docs/alerting/latest/alertmanager/#silences).

I think it necessary to notify operator authors of the usage on these two technologies where they will greatly improve the efficiency of alerting.
Also, not sure if it can be good to provide [a group manifest template](https://prometheus.io/docs/alerting/latest/configuration/#example) to focus on alert management over operators.

Author: @Kavinjsir  I'll be happy if you can add this part to the doc after it is merged.
I would consider this as an advanced topic.

* `runbook_url` annotation is a link to an alert runbook which is intended to guide a cluster owner and/or operator through the steps of fixing problems on clusters, which are surfaced by alerts.
  * An example for [Runbook style documentation](https://github.com/openshift/runbooks/blob/master/example.md).
  * Your operator alert runbooks can be saved at your operator repository, to `/docs/monitoring/runbooks/` for example,
    at [OpenShift Runbooks](https://github.com/openshift/runbooks) if your operator is shipped with OpenShift
    or another location that fits your operator.
  * If you are using Github, you can use [Github Pages](https://pages.github.com/) for a better view of the runbooks.
* `kubernetes_operator_part_of` label indicating the operator name. Label name is based on the  [Kubernetes Recommended Labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels).

#### Alerts Severity
##### Critical Alerts

For alerting current and impending disaster situations. These alerts

> page an SRE.  The situation should warrant waking someone in the middle of the

Reviewer 5: double spacing
```suggestion
page an SRE. The situation should warrant waking someone in the middle of the
```

they can respond with corrective action quickly.

Example critical alert: [KubeAPIDown](https://github.com/openshift/cluster-monitoring-operator/blob/79cdf68/assets/control-plane/prometheus-rule.yaml#L412-L421)

```yaml
- alert: KubeAPIDown
  annotations:
    summary: Target disappeared from Prometheus target discovery.
    description: KubeAPI has disappeared from Prometheus target discovery.
    runbook_url: https://github.com/openshift/runbooks/blob/master/alerts/cluster-monitoring-operator/KubeAPIDown.md
  expr: |
    absent(up{job="apiserver"} == 1)
  for: 15m
  labels:
    severity: critical
```

This alert fires if no Kubernetes API server instance has reported metrics
successfully in the last 15 minutes.  This is a clear example of a critical
control-plane issue that represents a threat to the operability of the cluster
as a whole, and likely warrants paging someone.  The alert has clear summary and
description annotations, and it links to a runbook with information on
investigating and resolving the issue.

The group of critical alerts should be small, very well defined, highly
documented, polished and with a high bar set for entry.

##### Warning Alerts

> The vast majority of alerts should use the severity.  Issues at the

Reviewer 5: ```suggestion
The vast majority of alerts should use the severity. Issues at the
```

      For more information refer to 'oc adm upgrade'{{ "{{ with $console_url := \"console_url\" | query }}{{ if ne (len (label \"url\" (first $console_url ) ) ) 0}} or {{ label \"url\" (first $console_url ) }}/settings/cluster/{{ end }}{{ end }}" }}.
    expr: |
      max by (name, condition, endpoint) (cluster_operator_conditions{name="version", condition="Upgradeable", endpoint="metrics"} == 0)
    for: 60m
    labels:
      severity: warning
```

This alert fires if one or more operators have not reported their `Upgradeable`
condition as true in more than an hour.  The alert has a clear name and
informative summary and description annotations.  The timeline is appropriate
for allowing the operator a chance to resolve the issue automatically, avoiding
the need to alert an administrator.

##### Info Alerts

Info level alerts represent situations an administrator should be aware

> of, but that don't necessarily require any action.  Use these sparingly, and

Reviewer 5: ```suggestion
of, but that don't necessarily require any action. Use these sparingly, and
```

consider instead reporting this information via Kubernetes events.

Example info alert: [MultipleContainersOOMKilled](https://github.com/openshift/cluster-monitoring-operator/blob/79cdf68/assets/cluster-monitoring-operator/prometheus-rule.yaml#L326-L338)

```yaml
- alert: MultipleContainersOOMKilled
  annotations:
    description: Multiple containers were out of memory killed within the past
      15 minutes. There are many potential causes of OOM errors, however issues
      on a specific node or containers breaching their limits is common.
      summary: Containers are being killed due to OOM
  expr: sum(max by(namespace, container, pod) (increase(kube_pod_container_status_restarts_total[12m]))
    and max by(namespace, container, pod) (kube_pod_container_status_last_terminated_reason{reason="OOMKilled"}) == 1) > 5
  for: 15m
  labels:
    namespace: kube-system
    severity: info
```

This alert fires if multiple containers have been terminated due to out of

> memory conditions in the last 15 minutes.  This is something the administrator

Reviewer 5: ```suggestion
memory conditions in the last 15 minutes. This is something the administrator
```
